lib/copy/ncp.js

```diff
@@ -65,13 +56,80 @@ function ncp (source, dest, options, callback) {
       }
 
       if (stats.isDirectory()) {
+        if (filter) return applyFilterOnDir()
         return onDir(item)
       } else if (stats.isFile() || stats.isCharacterDevice() || stats.isBlockDevice()) {
+        if (filter) return applyFilterOnFile()
         return onFile(item)
       } else if (stats.isSymbolicLink()) {
         // Symlinks don't really need to know about the mode.
         return onLink(source)
       }
+
+      function applyFilterOnDir () {
+        if (filter instanceof RegExp) {
+          console.warn('Warning: fs-extra: Passing a RegExp filter is deprecated, use a function')
+          if (!filter.test(source)) {
+            if (noRecurseOnFailedFilter) return doneOne()
+            return readFailedDir()
+          }
+          return mkParentDir()
+        } else if (typeof filter === 'function') {
+          if (!filter(source, dest)) {
+            if (noRecurseOnFailedFilter) return doneOne()
+            return readFailedDir()
+          }
+          return mkParentDir()
+        }
+
+        function readFailedDir () {
+          fs.readdir(source, (err, items) => {
+            if (err) return callback(err)
+            Promise.all(items.map(item => {
+              return new Promise((resolve, reject) => {
+                ncp(path.join(source, item), path.join(dest, item), options, err => {
+                  if (err) reject(err)
+                  else resolve()
+                })
+              })
+            })).then(() => callback()).catch(callback)
+          })
+        }
+
+        function mkParentDir () {
+          var destParent = path.dirname(dest)
+          pathExists(destParent, (err, dirExists) => {
+            if (err) return callback(err)
+            if (dirExists) return onDir(item)
+            mkdirp(destParent, err => {
```

Comment:
No need for this, `mkdirp` won't error if the directory already exists. Just call `mkdirp` without checking if it exists.